### PR TITLE
Add 2nd order heun scheduler

### PR DESCRIPTION
**Pull request by patrickvonplaten**
This PR does two things:

- 1. It adds the first second order scheduler to `diffusers` without changing the logic of the scheduler / model API:
  - a. The loop is still defined by iterating over `self.scheduler.timesteps`
  - b. The scheduler step function still returns only the computed model values
- 2. It fixes the progress bar of PNDM and also makes it compatible with 2nd order schedulers. Now PNDM shows 50 update steps instead of 51.

Note that this design might evolve in the future as discussed in https://github.com/huggingface/diffusers/issues/1308

### CODE REVIEW

File: src/diffusers/schedulers/scheduling_heun.py [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+        timesteps = np.linspace(0, num_train_timesteps - 1, num_inference_steps, dtype=float)[::-1].copy()
+
+        sigmas = np.array(((1 - self.alphas_cumprod) / self.alphas_cumprod) ** 0.5)
+        sigmas = np.interp(timesteps, np.arange(0, len(sigmas)), sigmas)
+        sigmas = np.concatenate([sigmas, [0.0]]).astype(np.float32)
+        sigmas = torch.from_numpy(sigmas).to(device=device)
+        self.sigmas = torch.cat([sigmas[:1], sigmas[1:-1].repeat_interleave(2), sigmas[-1:]])
+
+        timesteps = torch.from_numpy(timesteps)
+        timesteps = torch.cat([timesteps[:1], timesteps[1:].repeat_interleave(2), timesteps[-1:]])
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_heun.py:118]: @patil-suraj @pcuenca @anton-l - this just repeats timesteps for the second order

File: src/diffusers/schedulers/scheduling_heun.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+        num_train_timesteps = num_train_timesteps or self.config.num_train_timesteps
+
+        timesteps = np.linspace(0, num_train_timesteps - 1, num_inference_steps, dtype=float)[::-1].copy()
+
+        sigmas = np.array(((1 - self.alphas_cumprod) / self.alphas_cumprod) ** 0.5)
+        sigmas = np.interp(timesteps, np.arange(0, len(sigmas)), sigmas)
+        sigmas = np.concatenate([sigmas, [0.0]]).astype(np.float32)
+        sigmas = torch.from_numpy(sigmas).to(device=device)
+        self.sigmas = torch.cat([sigmas[:1], sigmas[1:-1].repeat_interleave(2), sigmas[-1:]])
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_heun.py:115]: @patil-suraj @pcuenca @anton-l - this just repeats sigmas for the second order

File: src/diffusers/schedulers/scheduling_heun.py [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+        else:
+            self.timesteps = timesteps.to(device=device)
+
+        # empty dt and derivative
+        self.prev_derivative = None
+        self.dt = None
+
+    @property
+    def state_in_first_order(self):
+        return self.dt is None
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_heun.py:132]: simple property defining the mode of the scheduler

File: src/diffusers/schedulers/scheduling_heun.py [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+            model_output (`torch.FloatTensor` or `np.ndarray`): direct output from learned diffusion model. timestep
+            (`int`): current discrete timestep in the diffusion chain. sample (`torch.FloatTensor` or `np.ndarray`):
+                current instance of sample being created by diffusion process.
+            return_dict (`bool`): option for returning tuple rather than SchedulerOutput class
+        Returns:
+            [`~schedulers.scheduling_utils.SchedulerOutput`] or `tuple`:
+            [`~schedulers.scheduling_utils.SchedulerOutput`] if `return_dict` is True, otherwise a `tuple`. When
+            returning a tuple, the first element is the sample tensor.
+        """
+        step_index = (self.timesteps == timestep).nonzero().item()
```
Comment (keturn) [inline src/diffusers/schedulers/scheduling_heun.py:154]: Does this need to be adjusted to account for duplicate entries in `timesteps`? because `==` will match more than one thing and `.item()` complains about that.

Then the result is ambiguous, but I guess we can take the first and then add one if  `not state_in_first_order`?

File: src/diffusers/schedulers/scheduling_heun.py [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+        # passing it to the model which requires a change in API
+        gamma = 0
+        sigma_hat = sigma * (gamma + 1)  # Note: sigma_hat == sigma for now
+
+        # 1. compute predicted original sample (x_0) from sigma-scaled predicted noise
+        pred_original_sample = sample - sigma_hat * model_output
+
+        if self.state_in_first_order:
+            # 2. Convert to an ODE derivative
+            derivative = (sample - pred_original_sample) / sigma_hat
```
Comment (keturn) [inline src/diffusers/schedulers/scheduling_heun.py:175]:
```py
pred_original_sample = sample - sigma_hat * model_output
derivative = (sample - pred_original_sample) / sigma_hat

# implies
derivative = (sample - sample - sigma_hat * model_output) / sigma_hat
derivative = (-sigma_hat * model_output) / sigma_hat
derivative = -model_output
```

**Review by pcuenca — COMMENTED** [review]
This is a great effort, thanks a lot! I find it much easier to discuss based on some code than on the ether, so it's very helpful.

I think it's a bit finicky to get it right with all the indexing and staff, but the end result looks really understandable. I could get what was going in a first pass even though I didn't realize there were some details a bit off until I looked more carefully. I think this is acceptable for the "write-once, read-many" approach we want to achieve here.

As an exercise, I'll try to find some time to build the alternative example with the scheduler returning a tuple that I proposed in the other thread, and see how it compares to this.

File: src/diffusers/schedulers/scheduling_heun.py [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+
+            # 3. Retrieve 1st order derivative
+            dt = self.dt
+
+            # free dt and derivative
+            # Note, this puts the scheduler in "first order mode"
+            self.prev_derivative = None
+            self.dt = None
+
+        prev_sample = model_output + derivative * self.dt
```
[the remark this answers is not in the document]
Comment (patil-suraj) [reply src/diffusers/schedulers/scheduling_heun.py:195]: good catch

Comment (patil-suraj) [inline src/diffusers/schedulers/scheduling_heun.py:195]: cf https://github.com/crowsonkb/k-diffusion/blob/master/k_diffusion/sampling.py#L110
```suggestion
        prev_sample = sample + derivative * self.dt
```

File: src/diffusers/schedulers/scheduling_heun.py (hunk 4 above)
Comment (anton-l) [reply src/diffusers/schedulers/scheduling_heun.py:154]: The hack to get the timestep index seems to become even harder to read and debug here. Should we maybe return the timestep index arguments at some point? (planting a thought for `1.0.0`)

File: src/diffusers/schedulers/scheduling_heun.py (hunk 5 above)
Comment (pcuenca) [reply src/diffusers/schedulers/scheduling_heun.py:175]: I fixed the implementation in #1356. I can bring the appropriate changes here.

File: src/diffusers/schedulers/scheduling_heun.py [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+        pred_original_sample = sample - sigma_hat * model_output
+
+        if self.state_in_first_order:
+            # 2. Convert to an ODE derivative
+            derivative = (sample - pred_original_sample) / sigma_hat
+            # 3. 1st order derivative
+            dt = sigma_next - sigma_hat
+
+            # store for 2nd order step
+            self.prev_derivative = derivative
```
Comment (pcuenca) [inline src/diffusers/schedulers/scheduling_heun.py:180]:
```suggestion
            self.prev_derivative = derivative
            self.sample = sample
```

Store `sample` because the 2nd order step has to compute `prev_sample` from the original sample of the 1st step.

File: src/diffusers/schedulers/scheduling_heun.py (hunk 6 above)
Comment (pcuenca) [inline src/diffusers/schedulers/scheduling_heun.py:195]:
```suggestion
        prev_sample = self.sample + derivative * dt
```

@patil-suraj's correction, but starting from the original `sample` used in the first step.

**Review by pcuenca — COMMENTED** [review]
Submitted changes to make it work (I think).

**Comment by patrickvonplaten** [comment]
Thanks a lot for the corrections @pcuenca @patil-suraj @keturn.

The PR as is now is functional and gives 1-to-1 the same results as k-diffusion:

```python
#!/usr/bin/env python3
from diffusers import DiffusionPipeline, StableDiffusionPipeline, HeunDiscreteScheduler
import torch

seed = 33

pipe = DiffusionPipeline.from_pretrained("CompVis/stable-diffusion-v1-4", custom_pipeline="sd_text2img_k_diffusion")
pipe = pipe.to("cuda")

prompt = "an astronaut riding a horse on mars"
pipe.set_sampler("sample_heun")
generator = torch.Generator(device="cuda").manual_seed(seed)
image = pipe(prompt, generator=generator, num_inference_steps=20).images[0]

image.save("./astronaut_heun_k_diffusion_comp.png")

pipe = StableDiffusionPipeline(**pipe.components)
pipe = pipe.to("cuda")
pipe.scheduler = HeunDiscreteScheduler.from_config(pipe.scheduler.config)
generator = torch.Generator(device="cuda").manual_seed(seed)
image = pipe(prompt, generator=generator, num_inference_steps=20).images[0]

image.save("./astronaut_heun_comp.png")
```

**K Diffusion**
astronaut_heun_k_diffusion_comp

**This PR (Diffusers)**
astronaut_heun_comp (1)

File: src/diffusers/schedulers/scheduling_heun.py [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+        step_index: Union[int, torch.IntTensor],
+    ) -> torch.FloatTensor:
+        """
+        Args:
+        Ensures interchangeability with schedulers that need to scale the denoising model input depending on the
+        current timestep.
+            sample (`torch.FloatTensor`): input sample timestep (`int`, optional): current timestep
+        Returns:
+            `torch.FloatTensor`: scaled input sample
+        """
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply src/diffusers/schedulers/scheduling_heun.py:92]: Sorry yeah this was a bug

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 9]
```diff
@@ -540,30 +540,34 @@ def __call__(
...
+                if (i + 1) > num_warmup_steps and (i + 1) % self.scheduler.order == 0:
+                    progress_bar.update()
+                    if callback is not None and i % callback_steps == 0:
+                        callback(i, t, latents)
 
         # 8. Post-processing
         image = self.decode_latents(latents)
 
-        # 9. Run safety checker
+        # 9. Run safety checker as
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:570]:
```suggestion
        # 9. Run safety checker
```
